holoviews/plotting/plot.py

```diff
@@ -979,10 +979,21 @@ def _construct_callbacks(self):
         registry = list(Stream.registry.items())
         callbacks = Stream._callbacks[self.backend]
         for source in self.link_sources:
-            streams = [
-                s for src, streams in registry for s in streams
-                if src is source or (src._plot_id is not None and
-                                     src._plot_id == source._plot_id)]
+            streams = []
+            for src, src_streams in registry:
+                # Skip if source identities do not match
+                if (src is not source and (src._plot_id is None or src._plot_id != source._plot_id)):
+                    continue
+                for stream in src_streams:
+                    # Skip if Stream.source is an overlay but the plot isn't
+                    # or if the source is an element but the plot isn't
+                    src_el = stream.source.last if isinstance(stream.source, HoloMap) else stream.source
+                    if ((isinstance(src_el, CompositeOverlay) and
+                        not isinstance(self, GenericOverlayPlot)) or
+                        (isinstance(src_el, Element) and
+                         isinstance(self, GenericOverlayPlot))):
```

Comment:
```suggestion
                    if (
                        (isinstance(src_el, CompositeOverlay) and not isinstance(self, GenericOverlayPlot)) 
                        or (isinstance(src_el, Element) and isinstance(self, GenericOverlayPlot))
                    ):
```

Easier for me to read.